### PR TITLE
Stylus support linux

### PR DESCRIPTION
**Pull request by karavindhan**
*Adding stylus input pointer support for linux based on GdkEvent pressure axis*

*https://github.com/flutter/flutter/issues/63209*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by moffatman** [comment]
Needs a test, it isn't easy to mock these GdkEvents but I have an example of doing so for testing trackpad scrolling:

https://github.com/flutter/engine/blob/main/shell/platform/linux/fl_scrolling_manager_test.cc#L224-L248

File: shell/platform/embedder/embedder.h [hunk 1]
```diff
@@ -926,6 +926,8 @@ typedef struct {
   double scale;
   /// The rotation of the pan/zoom in radians, where 0.0 is the initial angle.
   double rotation;
+  /// The pen pressure from stylus between 0.0 and 1.0
```
Comment (robert-ancell) [inline shell/platform/embedder/embedder.h:929]: Missing '.' at the end of the comment.

File: shell/platform/linux/fl_engine.cc [hunk 2]
```diff
@@ -176,9 +177,9 @@ static bool compositor_present_layers_callback(const FlutterLayer** layers,
                                                void* user_data) {
   g_return_val_if_fail(FL_IS_RENDERER(user_data), false);
   return fl_renderer_present_layers(FL_RENDERER(user_data), layers,
+
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:180]: Accidental whitespace change?

File: shell/platform/linux/fl_engine.cc [hunk 3]
```diff
@@ -176,9 +177,9 @@ static bool compositor_present_layers_callback(const FlutterLayer** layers,
...
                                     layers_count);
 }
-
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:181]: Accidental whitespace removal?

File: shell/platform/linux/fl_view.cc [hunk 4]
```diff
@@ -179,18 +182,39 @@ static void check_pointer_inside(FlView* view, GdkEvent* event) {
...
+          view->engine, kAdd, kind,
           gdk_event_get_time(event) * kMicrosecondsPerMillisecond,
-          x * scale_factor, y * scale_factor, 0, 0, view->button_state);
+          x * scale_factor, y * scale_factor, 0, 0, view->button_state,
+          pressure);
     }
   }
 }
 
+FlutterPointerDeviceKind check_device_is_stylus(FlView* view,
```
Comment (robert-ancell) [inline shell/platform/linux/fl_view.cc:199]: Should be `Flview *self`. Function should be called fl_view_check_device_is_stylus as it's a method.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Other than the code style fixes looks good.

File: shell/platform/linux/fl_engine.cc (hunk 3 above)
Comment (karavindhan) [reply shell/platform/linux/fl_engine.cc:181]: Yes. Something with my vscode.

File: shell/platform/linux/fl_view.cc (hunk 4 above)
Comment (karavindhan) [reply shell/platform/linux/fl_view.cc]: Renamed.

**Comment by karavindhan** [comment]
> Needs a test, it isn't easy to mock these GdkEvents but I have an example of doing so for testing trackpad scrolling:
> 
> https://github.com/flutter/engine/blob/46ea2b3cce1f25a91ed8bc9db150bd0add3d0f2e/shell/platform/linux/fl_scrolling_manager_test.cc#L224-L249

@moffatman Thanks for the pointer, totally with you as GdkDevice structures are not exported.  For my case to write test cases that is meaningful I have to write a new mocked GdkDeviceClass that implements get_state method. Only possibility is write mock device add it to upstream gdk. 
Still confused on how to go about this.

File: shell/platform/linux/fl_engine.cc (hunk 3 above)
Comment (dkwingsmt) [reply shell/platform/linux/fl_engine.cc:181]: Then shall we revert it?

**Comment by dkwingsmt** [comment]
On mocking `GdkEvent`: I suggest creating a new struct called, for example, `FlPointerEvent`, which is an exact replica of the information of `GdkEvent` that we need. The first Flutter callback that receives `GdkEvent` converts it to `FlPointerEvent`, and all internal functions and unit tests only use `FlPointerEvent`. 

The keyboard module is already using this technique: https://github.com/flutter/engine/blob/main/shell/platform/linux/fl_key_event.h#L31

**Comment by dkwingsmt** [comment]
Regarding unit tests: It seems that there are no existing tests for `FlView` at all (which is surprising). I wonder if you can just create an `FlView` in unit tests or you need any mocking. Let me know if you meet any obstacles. `FlEngine` can already be mocked using `make_mock_engine`.

Other than that, you can find many existing tests as a starting point. For example, https://github.com/flutter/engine/blob/main/shell/platform/linux/fl_accessible_text_field_test.cc#L148 shows you how to "hijack" embedder APIs, which you can use to verify `SendPointerEvent` calls.

File: shell/platform/linux/fl_view_private.h [hunk 5]
```diff
@@ -52,4 +54,8 @@ void fl_view_add_widget(FlView* view,
  */
 void fl_view_end_frame(FlView* view);
 
+FlutterPointerDeviceKind fl_view_check_device_is_stylus(FlView* self,
```
Comment (robert-ancell) [inline shell/platform/linux/fl_view_private.h:57]: This needs a comment above it describing the function.

File: shell/platform/linux/fl_engine.cc [hunk 6]
```diff
@@ -178,7 +179,6 @@ static bool compositor_present_layers_callback(const FlutterLayer** layers,
   return fl_renderer_present_layers(FL_RENDERER(user_data), layers,
                                     layers_count);
 }
-
```
Comment (robert-ancell) [inline shell/platform/linux/fl_engine.cc:181]: One more whitespace removal that shouldn't be there.

File: shell/platform/linux/fl_view_private.h (hunk 5 above)
Comment (robert-ancell) [inline shell/platform/linux/fl_view_private.h:57]: This is not used outside fl_view.c, so doesn't need to be exported in a header.

File: shell/platform/linux/fl_view.cc [hunk 7]
```diff
@@ -166,10 +166,13 @@ static gboolean send_pointer_button_event(FlView* self, GdkEventButton* event) {
                                                event->y * scale_factor);
   fl_keyboard_manager_sync_modifier_if_needed(self->keyboard_manager,
                                               event->state, event->time);
-  fl_engine_send_mouse_pointer_event(
-      self->engine, phase, event->time * kMicrosecondsPerMillisecond,
-      event->x * scale_factor, event->y * scale_factor, 0, 0,
-      self->button_state);
+  gdouble pressure;
+  FlutterPointerDeviceKind kind =
```
[the remark this answers is not in the document]
Comment (robert-ancell) [reply shell/platform/linux/fl_view.cc:170]: I'd also move `fl_engine_send_pointer_event` into `fl_view.cc` - there's no other code that needs to call this function outside of there.

What I'd expect to see is:
```c
static void send_pointer_event(FlView *self, GdkEvent *event) {
  // Logic to determine if it's a stylus event and get pressure.
  // Call fl_engine_send_stylus_pointer_event or fl_engine_send_mouse_pointer_event based on the above.
}
```

[the remark this answers is not in the document]
Comment (karavindhan) [reply shell/platform/linux/fl_view.cc:170]: Agreed, will make those changes.

File: shell/platform/linux/fl_view_private.h (hunk 5 above)
Comment (karavindhan) [reply shell/platform/linux/fl_view_private.h:57]: Got it.

**Comment by karavindhan** [comment]
> Regarding unit tests: It seems that there are no existing tests for `FlView` at all (which is surprising). I wonder if you can just create an `FlView` in unit tests or you need any mocking. `FlEngine` can already be mocked using `make_mock_engine`.
> 
> Other than that, you can find many existing tests as a starting point. For example, https://github.com/flutter/engine/blob/main/shell/platform/linux/fl_accessible_text_field_test.cc#L148 shows you how to "hijack" embedder APIs, which you can use to verify `SendPointerEvent` calls.
> 
> Let me know if you find any other obstacles.

Adapted your advise and implemented tests and but when I run them I get a simple header include error and I can't decipher why it happens,
`../../flutter/shell/platform/linux/fl_view_test.cc:46:10: error: variable has incomplete type 'FlView' (aka '_FlView')
  FlView view;
         ^
../../flutter/shell/platform/linux/public/flutter_linux/fl_view.h:19:1: note: forward declaration of '_FlView'
G_DECLARE_FINAL_TYPE(FlView, fl_view, FL, VIEW, GtkContainer)
^
../../build/linux/debian_sid_amd64-sysroot/usr/include/glib-2.0/gobject/gtype.h:1405:18: note: expanded from macro 'G_DECLARE_FINAL_TYPE'
  typedef struct _##ModuleObjName ModuleObjName;                                                         \
                 ^
<scratch space>:122:1: note: expanded from here
_FlView
^
../../flutter/shell/platform/linux/fl_view_test.cc:93:10: error: variable has incomplete type 'FlView' (aka '_FlView')
  FlView view;
         ^
../../flutter/shell/platform/linux/public/flutter_linux/fl_view.h:19:1: note: forward declaration of '_FlView'
G_DECLARE_FINAL_TYPE(FlView, fl_view, FL, VIEW, GtkContainer)
^
../../build/linux/debian_sid_amd64-sysroot/usr/include/glib-2.0/gobject/gtype.h:1405:18: note: expanded from macro 'G_DECLARE_FINAL_TYPE'
  typedef struct _##ModuleObjName ModuleObjName;                                                         \
                 ^
<scratch space>:122:1: note: expanded from here
_FlView
^
2 errors generated.
[18/37] LINK ./ui_unittests
ninja: build stopped: subcommand failed`

Might be I am missing something simple, :)

**Comment by dkwingsmt** [comment]
@karavindhan I'm deeply sorry that this PR was slipped through my mind and I had been completely missing your question.

I tried to update your code and made a compilable result at https://github.com/flutter/engine/pull/42270. The major changes are:
* Updated it with the latest main (there were a few merge conflicts, nothing major though.)
* Added your new files to license and BUILD.gn.
* Resolved your problem by using proper initializing functions for the classes.

Now, the problem with your approach is that you've been trying to directly create a `FlView` class, but that's not how GObject (the OO framework that GDK and the Flutter Linux embedder use) works. In GObject, most classes, like `FlView`, only exposes a partial definition and a initializing function (such as `fl_view_new`), meaning that the only way to create them is through the initializing function and you can only ever get an opaque pointer. You can not get the implementation of the struct.

I've fixed your code by adding a few private functions that are only available to unit tests. However I haven't completed the fix since the second unit test of yours probably need to mock `GdkDevice` in order to allow the `fl_pointer_check_device_is_stylus` function to read "stylus". (Also I can't see the meaning of the last few lines of this function, I guess you haven't finished it yet?)

I understand that it's been such a long time and I again sincerely apologize. If you still wish to move forward, let me know! Also, it might be easier for you if I create the `FlPointerEvent` and its unit tests first so that you can modify based on that. If you prefer this way, also let me know. :)

**Comment by karavindhan** [comment]
@dkwingsmt No issues at at all. Since flutter build system is flexible enough, I have been using a local build of Flutter with stylus support for building digital painting app.  I think I didn't put much investment into learning GObject stuff, my bad, thanks a lot for fixing things up.  I would be happy to have this mainstream as well, I had across some issues with local build and I will test against your branch to see whether they still have to be added here as well before start of next week.
`Also, it might be easier for you if I submit a patch to create the FlPointerEvent and its unit tests first so that you can modify based on that. ` It sounds good to me as well

**Comment by CodeDoctorDE** [comment]
Hi, is there anything blocked here?
#42270 was last updated on May 24.
It would be nice if I can know the current status.

**Comment by dkwingsmt** [comment]
I worked a bit on that issue but I really don't have the bandwidth for it at the moment. But I definitely plan to come back once I catch a breath. Sorry for the delay.

**Comment by dkwingsmt** [comment]
I'm closing this PR for now. I still have this on my mind but I can't find bandwitdh to work on it. As soon as I'm almost done with multi-view I'll come back, unless someone else picks it up.

**Comment by BreakingLead** [comment]
Is there any progress?

**Pull request closed without merging**